rfcs/0011-reusable-reactors-folder.md

```diff
@@ -0,0 +1,115 @@
+
+# Proposed Implementation
+[proposed-implementation]: #proposed-implementation
+
+### New project Structure
+The current structure of a LF project is organized as follows:
+
+```shell
+├── root
+│   ├── bin/       # Binary executables
+│   ├── include/   # Header files
+│   ├── src/       # LF programs
+│   ├── src-gen/   # Generated code for local LF programs
+│   └── fed-gen/   # Generated code for federated programs (instead of src-gen)
+```
+
+To enhance the project structure, the following additions are proposed:
+
+- **`lib/`**: Directory for storing reusable reactors.
+- **`Lingo.toml`**: Configuration file crucial for project sharing, storing Lingo configuration details including external LF projects to include (refer to [Lingo Documentation](https://github.com/lf-lang/lingo?tab=readme-ov-file#the-toml-based-package-configurations) for further details). The internal structure of this file will be discussed in another RFC.
+
+The refined structure will appear as follows:
+```shell
+├── root
+│   ├── bin/
+│   ├── include/
+│   ├── lib/ # Directory for storing reusable reactors
+│   │   ├── Input.lf # Ex: reactor capturing external inputs (e.g., Microphone, Camera)
+│   │   └── ComputerVision.lf # Ex: reactor performing computer vision tasks (e.g., object detection, face recognition)
+│   ├── src/
+│   ├── src-gen/
+│   ├── fed-gen/ # Only for federated programs (instead of src-gen)
+│   └── Lingo.toml # Configuration file for Lingo Package Manager
+```
+The `lib/` directory is intended to store only reusable reactors, allowing them to be shared across multiple projects and reducing the need for reimplementation. The `src/` folder, on the other hand, is for executable programs that might use reactors defined in the `lib/` folder to develop a program. Generally, `src/` should contain only files that the developer compiles and executes. Note that the content of directories such as `src-gen`, `fed-gen`, and `include` are outside the scope of this RFC.
+
+This structured approach promotes efficient code reuse, simplifies integration of external resources, and fosters collaborative development practices within the LF community.
+
+### Supporting reactors
+[supporting-reactors]: #supporting-reactors
+
+According to this proposal, the `lib/` directory is designated for reusable reactors, while the `src/` directory is reserved for executable programs. In some projects, there may be reactors that are used internally by a library but are not intended to be exposed to its users. These can be considered "supporting reactors." To clearly mark such reactors as internal and not for public use, we propose using a `private/` directory within the `lib/` folder. This directory can exist at any level of hierarchy within `lib/` and signifies that its contents are private and should not be exposed to users.
```

Comment:
> internal and not for public use, we propose using a `private/` directory within the `lib/` folder. This directory can exist at any level of hierarchy within `lib/` and signifies that its contents are private and should not be exposed to users.

What is meant with "public use" here, and what is a "user"? Wouldn't it suffice to just not put the file in `lib` if it's not meant to be exposed? It seems unusual to work a scoping mechanism into the filesystem layout. Is there any precedent for this in other language, or did we come up with this ourselves?

Reply:
I personally see it as a hack. 

Because the LF language doesn't have any concept of visibility or modules.

Reply:
We discussed this during the Monday meeting and it became clear to me that the intent here was to propose a convention, and not to implement or enforce any checks. Please revise the text to clarify this intent.